dask_planner/src/sql/optimizer/join_reorder.rs

```diff
@@ -0,0 +1,385 @@
+
+    // Merge both the lists of dimensions by giving user order
+    // the preference for tables without a selective predicate,
+    // whereas for tables with selective predicates giving preference
+    // to smaller tables. When comparing the top of both
+    // the lists, if size of the top table in the selective predicate
+    // list is smaller than top of the other list, choose it otherwise
+    // vice-versa.
+    // This algorithm is a greedy approach where smaller
+    // joins with filtered dimension table are preferred for execution
+    // earlier than other Joins to improve Join performance. We try to keep
+    // the user order intact when unsure about reordering to make sure
+    // regressions are minimized.
+    let mut result = vec![];
+    while !filtered_dimensions.is_empty() || !unfiltered_dimensions.is_empty() {
+        if !filtered_dimensions.is_empty() && !unfiltered_dimensions.is_empty() {
+            if filtered_dimensions[0].size < unfiltered_dimensions[0].size {
+                result.push(filtered_dimensions.remove(0));
+            } else {
+                result.push(unfiltered_dimensions.remove(0));
+            }
+        } else if !filtered_dimensions.is_empty() {
+            result.push(filtered_dimensions.remove(0));
+        } else {
+            result.push(unfiltered_dimensions.remove(0));
+        }
+    }
+    assert!(filtered_dimensions.is_empty());
+    assert!(unfiltered_dimensions.is_empty());
+
+    let dim_plans: Vec<LogicalPlan> = result.iter().map(|rel| rel.plan.clone()).collect();
+
+    let mut join_conds = HashSet::new();
+    for cond in &conds {
+        match cond {
+            (Expr::Column(l), Expr::Column(r)) => {
+                join_conds.insert((l.clone(), r.clone()));
+            }
+            _ => {
+                return Ok(Some(plan.clone()));
+            }
+        }
+    }
+
+    let optimized = if facts.len() == 1 {
+        build_join_tree(&facts[0].plan, &dim_plans, &mut join_conds)?
+    } else {
+        // Build one join tree for each fact table
+        let fact_dim_joins = facts
+            .iter()
+            .map(|f| build_join_tree(&f.plan, &dim_plans, &mut join_conds))
+            .collect::<Result<Vec<_>>>()?;
+        // Join the trees together
+        build_join_tree(&fact_dim_joins[0], &fact_dim_joins[1..], &mut join_conds)?
+    };
+
+    if join_conds.is_empty() {
+        Ok(Some(optimized))
+    } else {
+        Ok(Some(plan.clone()))
+    }
+}
+
+/// Represents a Fact or Dimension table, possibly nested in a filter
+#[derive(Clone, Debug)]
+struct Relation {
+    /// Plan containing the table scan for the fact or dimension table
+    /// May also contain Filter and SubqueryAlias
+    plan: LogicalPlan,
+    /// Estimated size of the underlying table before any filtering is applied
+    size: usize,
+}
+
+impl Relation {
+    fn new(plan: LogicalPlan) -> Self {
+        let size = get_table_size(&plan).unwrap_or(100);
```

Comment:
I feel like a warning should be issued here if we default to 100. I could see situations where undesired optimizing could occur if this defaulted to 100. This isn't something we can "fix" but we should alert the user that it is happening so they can be aware.

Reply:
Thanks for reviewing! Updated.